dali/python/nvidia/dali/plugin/pytorch/experimental/proxy/__init__.py

```diff
@@ -0,0 +1,771 @@
+                _nvtx.range_push("feed_pipeline")
+                for pipe_run_ref in batches:
+                    for name, data in pipe_run_ref.call_args.items():
+                        self._pipe.feed_input(name, data)
+                    self._pipe._run_once()
+                    fed_batches.append(pipe_run_ref.batch_id)
+                _nvtx.range_pop()
+
+            # If no batches to consume, continue
+            if len(fed_batches) == 0:
+                continue
+
+            _nvtx.range_push("outputs")
+            batch_id = fed_batches.pop(0)  # we are sure there's at least one
+            err = None
+            torch_outputs = None
+            try:
+                pipe_outputs = self._pipe.outputs()
+                torch_outputs = tuple(
+                    [
+                        to_torch_tensor(out.as_tensor(), not self._pipe.exec_dynamic)
+                        for out in pipe_outputs
+                    ]
+                )
+            except Exception as exception:
+                err = exception
+
+            self._dali_output_q.put((batch_id, torch_outputs, err))
+            _nvtx.range_pop()
+
+    def start_thread(self):
+        """
+        Starts the DALI pipeline thread. Note: Using scope's __enter__/__exit__ is preferred
+        """
+        if self._thread is not None:
+            return
+        self._thread = threading.Thread(target=DALIServer._thread_fn, args=(self,))
+        self._thread_stop_event = threading.Event()
+        self._thread.start()
+
+    def stop_thread(self):
+        """
+        Stops the DALI pipeline thread. Note: Using scope's __enter__/__exit__ is preferred
+        """
+        if self._thread_stop_event is None:
+            return
+        self._thread_stop_event.set()
+        self._thread.join()
+        self._thread = None
+        self._thread_stop_event = None
+
+    def __enter__(self):
+        """
+        Starts the DALI pipeline thread
+        """
+        self.start_thread()
+        return self
+
+    def __exit__(self, exc_type, exc_value, tb):
+        """
+        Stops the DALI pipeline thread
+        """
+        self.stop_thread()
+        if exc_type is not None:
+            warnings.warn(f"An exception occurred: {exc_value}", category=UserWarning)
+        return False  # Return False to propagate exceptions
+
+
+class DataLoader(_torchdata.dataloader.DataLoader):
+    """
+    DALI data loader to be used in the main loop, which replaces the pipeline run references
+    with actual data produced by the DALI server.
+    See :class:`nvidia.dali.plugin.pytorch.experimental.proxy.DALIServer` for a full example.
+    """
+
+    class _Iter(_torchdata.dataloader._MultiProcessingDataLoaderIter):
+        """
+        Data loader iterator used by the DALI proxy data loader
+        """
+
+        def __init__(self, loader):
+            super().__init__(loader)
+            self.loader = loader
+
```

Comment:
I guess it would be good to add `__del__` with stop_thread - in case the user breaks the loader loop somehow in the middle.

Reply:
added __del__ to loader